File: web/platform/purecart.mjs
// purecart.mjs — the cart that isn't one (DESIGN.md §12, the infeasible cart).
//
// `createRuntime` wants a cart: `init/tick/draw`. This supplies one that
// contains no game. It reads input, asks the scene what is under the pointer,
// submits the command the world offered, and draws what the world concluded.
// Point it at a different `.story` and it is a different game with no edit
// here — which is the whole claim, and the reason this file is short.
//
// What is left in JS after this is exactly the residue worth naming: the loop,
// the layout arithmetic inside the scene, and the ASSETS. Assets are not code —
// a sprite atlas is pixels, and pixels are not expressible as rules for the
// same reason a `.png` is not.

import { createScene } from './scene.mjs';

const verb = (term) => term.replace(/\(.*/, '');

/**
 * @param {object} o
 * @param {object} o.world   the generated binding session
 * @param {object} o.iface   its IFACE export
 * @param {object} o.doms    { ...SORTS, ...ENUMS }
 * @param {object} o.sheets  the atlases, by name — assets, not logic
 */
export function pureCart({ world, iface, doms, sheets, resolution = [640, 360] }) {
  const scene = createScene(world, iface, doms);

  return {
    resolution,
    sheets,

    init(ctx) {
      this.scene = scene;
      this.why = '';
      this.fx = [];
      ctx.why = '';
      scene.rebuild();
    },

    tick(ctx) {
      const { input } = ctx;
      if (!input.pressed(0)) return null;
      const target = scene.hit(input.pointer());
      if (!target) { this.why = ctx.why = ''; return null; }

      if (target.kind === 'entity') {
        this.why = ctx.why = '';
        const act = scene.pick(target.entity);
        return act ? ctx.world.actions().add(act) : null;
      }
      if (target.ok) {
        this.why = ctx.why = '';
        // the ground action IS the term the engine named; nothing to bind
        return ctx.world.actions().add({ action: verb(target.term),
                                         args: {}, term: target.term });
      }
      // A refused command explains itself, and the literal to explain is the
      // GUARD the engine says refused it — the world's own argument, not a
      // message this file invented or a judgment the story wrote to mirror it.
      this.why = ctx.why = target.blockers?.length
        ? ctx.world.why(target.blockers[0].atom, target.blockers[0].neg) : '';
      return null;
    },

    after(ctx) {
      // Cues are a LOOKUP, not a handler per event: the story already said
      // which sound a cue plays and which word it floats, so this reads the
      // table it declared. An emission `heave(guard)` is the cue `q_heave`.
      const sounds = scene.pairs('cue_sound'), words = scene.pairs('cue_word');
      for (const emitted of ctx.step.emits) {
        const q = `q_${emitted.replace(/\(.*/, '')}`;
        for (const [c, snd] of sounds) if (c === q) ctx.audio.sound(snd, 0.6);
        for (const [c, wrd] of words) if (c === q)
          this.fx.push({ text: scene.say(wrd), life: 40 });
      }
      scene.rebuild();
    },

    draw(ctx) {
      ctx.why = this.why;
      scene.draw(ctx, 'main');
      let i = 0;
      for (const f of this.fx) {
        ctx.draw.print(f.text, 300, 200 - (40 - f.life) / 2 - i * 10, 6, { big: true });
        f.life--; i++;
      }
      this.fx = this.fx.filter((f) => f.life > 0);
    },
  };
}

File: src/runtime/purecart.c
#include "runtime/purecart.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum { PC_MAXFX = 8, PC_WHY = 4096, PC_MAXTARGETS = PLAT_MAX_TARGETS };

typedef struct { char text[48]; int life; } floater;

struct purecart {
    world  *w;
    intern *syms;
    const iface *f;
    plat   *p;
    scene  *sc;
    char    sheet[32];
    char    why[PC_WHY];
    floater fx[PC_MAXFX];
    int     nfx;
};

purecart *purecart_new(world *w, intern *syms, const iface *f, plat *p,
                       const char *sheet)
{
    purecart *c = calloc(1, sizeof *c);
    c->w = w; c->syms = syms; c->f = f; c->p = p;
    c->sc = scene_new(w, syms, f, p);
    snprintf(c->sheet, sizeof c->sheet, "%s", sheet ? sheet : "main");
    return c;
}

void purecart_free(purecart *c)
{
    if (!c) return;
    scene_free(c->sc);
    free(c);
}

scene *purecart_scene(purecart *c) { return c->sc; }
const char *purecart_why(const purecart *c) { return c->why; }

/* The why-trace as text. `world_why` writes to a FILE*, and a runtime that
 * wants the string takes the standard-C route rather than a POSIX memory
 * stream — the native player has to build where `open_memstream` does not
 * exist. */
static void why_string(purecart *c, dl_lit q)
{
    c->why[0] = 0;
    FILE *tmp = tmpfile();
    if (!tmp) return;
    world_why(c->w, q, tmp);
    fflush(tmp);
    rewind(tmp);
    size_t n = fread(c->why, 1, sizeof c->why - 1, tmp);
    c->why[n] = 0;
    fclose(tmp);
}

static void pc_init(void *ctx, rt *r)
{
    purecart *c = ctx;
    (void)r;
    scene_rebuild(c->sc);
}

static int pc_tick(void *ctx, rt *r, uint32_t *out, int cap)
{
    purecart *c = ctx;
    (void)r;
    if (cap < 1) return 0;

    /* The PORTABLE input model (§12): declare what is focusable, ask what was
     * confirmed. This driver never reads a pointer or a key, so it plays
     * identically on a mouse, a touchscreen and a d-pad — and a cart that
     * reached for the pointer here would have been a desktop-only cart with
     * nothing saying so. */
    plat_target targets[PC_MAXTARGETS];
    int n = scene_targets(c->sc, targets, PC_MAXTARGETS);
    plat_focus_targets(c->p, targets, n);
    const char *id = plat_focus_confirmed(c->p);
    if (!id) return 0;

    scene_hit hit;
    if (!scene_target(c->sc, id, &hit)) { c->why[0] = 0; return 0; }

    if (hit.kind == SCENE_ENTITY) {
        c->why[0] = 0;
        const char *term = scene_pick_action(c->sc, hit.entity);
        if (!term) return 0;
        out[0] = intern_id(c->syms, term);
        return 1;
    }
    if (hit.ok) {
        c->why[0] = 0;
        /* the ground action IS the term the engine named; nothing to bind */
        out[0] = intern_id(c->syms, hit.term);
        return 1;
    }
    /* A refused command explains itself, and the literal to explain is the
     * GUARD the engine says refused it — the world's own argument, not a
     * message this file invented or a judgment the story wrote to mirror it. */
    if (hit.has_blocker) why_string(c, hit.blocker);
    else c->why[0] = 0;
    return 0;
}

static void pc_after(void *ctx, rt *r)
{
    purecart *c = ctx;
    /* Cues are a LOOKUP, not a handler per event: the story already said which
     * sound a cue plays and which word it floats, so this reads the table it
     * declared. An emission `heave(guard)` is the cue `q_heave`. */
    scene_pair sounds[64], words[64];
    int ns = scene_pairs(c->sc, "cue_sound", sounds, 64);
    int nw = scene_pairs(c->sc, "cue_word", words, 64);

    int nem = 0;
    const uint32_t *em = world_emits(rt_world(r), &nem);
    for (int i = 0; i < nem; i++) {
        const char *name = intern_name(c->syms, em[i]);
        char cue[64];
        size_t k = 0;
        cue[k++] = 'q'; cue[k++] = '_';
        for (size_t j = 0; name[j] && name[j] != '(' && k + 1 < sizeof cue; j++)
            cue[k++] = name[j];
        cue[k] = 0;
        for (int j = 0; j < ns; j++)
            if (strcmp(sounds[j].a, cue) == 0) plat_sound(c->p, sounds[j].b, 0.6f);
        for (int j = 0; j < nw; j++) {
            if (strcmp(words[j].a, cue) != 0) continue;
            if (c->nfx >= PC_MAXFX) continue;
            scene_say(words[j].b, c->fx[c->nfx].text, sizeof c->fx[0].text);
            c->fx[c->nfx].life = 40;
            c->nfx++;
        }
    }
    scene_rebuild(c->sc);
}

static void pc_draw(void *ctx, rt *r)
{
    purecart *c = ctx;
    (void)r;
    scene_draw(c->sc, c->sheet, c->why);
    int shown = 0;
    for (int i = 0; i < c->nfx; i++) {
        plat_print(c->p, c->fx[i].text, 300,
                   200 - (40 - c->fx[i].life) / 2 - shown * 10, 6, true);
        c->fx[i].life--;
        shown++;
    }
    int keep = 0;
    for (int i = 0; i < c->nfx; i++)
        if (c->fx[i].life > 0) c->fx[keep++] = c->fx[i];
    c->nfx = keep;
}

rt_cart purecart_cart(purecart *c)
{
    rt_cart cart;
    memset(&cart, 0, sizeof cart);
    cart.ctx = c;
    cart.init = pc_init;
    cart.tick = pc_tick;
    cart.draw = pc_draw;
    cart.after = pc_after;
    return cart;
}
